### PR TITLE
Improve logic for reading page column indexes in parquet

### PR DESCRIPTION
**Pull request by raunaqmorarka**
Prefetch the required portions of file for column/offset index
instead of making multiple small requests.
Lazily read offset index only when they're required.

### CODE REVIEW

**Comment by raunaqmorarka** [comment]
Encountered test failure due to https://github.com/trinodb/trino/issues/9432

**Comment by raunaqmorarka** [comment]
Encountered test failure due to https://github.com/trinodb/trino/issues/9433

**Comment by JamesRTaylor** [comment]
Is this different than the batching that occurs as a result of [this](https://github.com/trinodb/trino/blob/4035055d9cd2f6771ed485c2f8ad6daeec209821/lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java#L203) call in ParquetReader?

**Comment by raunaqmorarka** [comment]
> Is this different than the batching that occurs as a result of [this](https://github.com/trinodb/trino/blob/4035055d9cd2f6771ed485c2f8ad6daeec209821/lib/trino-parquet/src/main/java/io/trino/parquet/reader/ParquetReader.java#L203) call in ParquetReader?

I think that batching is about reading of the actual data. This one is about the reads done for the metadata (column and offset index).

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java [hunk 1]
```diff
@@ -24,167 +25,202 @@
...
+    private boolean columnIndexRead;
+    private boolean offsetIndexRead;
+
+    /**
+     * Creates a column index store which lazily reads column/offset indexes for the columns in paths.
+     *
+     * @param columnsRead is the set of columns used for projection
+     * @param columnsFiltered is the set of columns used for filtering
+     */
+    public static ColumnIndexStore create(
```
Comment (skrzypo987) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:65]: Is there a reason to use a static creator instead of just exposing the constructor?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java [hunk 2]
```diff
@@ -24,167 +25,202 @@
...
-            OffsetIndex offsetIndex;
-            try {
-                offsetIndex = readOffsetIndex(meta);
-            }
-            catch (IOException e) {
-                // If the I/O issue still stands it will fail the reading later;
-                // otherwise we fail the filtering only with a missing offset index.
-                offsetIndex = null;
+        this.dataSource = dataSource;
+        this.columnIndexReferences = new ArrayList<>(columnsFiltered.size());
```
Comment (skrzypo987) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:81]: Make those immutable at the end

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java [hunk 3]
```diff
@@ -24,167 +25,202 @@
...
+    private static final Logger log = Logger.get(TrinoColumnIndexStore.class);
 
-        OffsetIndex getOffsetIndex();
+    private final ParquetDataSource dataSource;
+    private final List<ColumnIndexMetadata> columnIndexReferences;
+    private final List<ColumnIndexMetadata> offsetIndexReferences;
+
+    private Map<ColumnPath, ColumnIndex> columnIndexStore;
+    private Map<ColumnPath, OffsetIndex> offsetIndexStore;
+    private boolean columnIndexRead;
```
Comment (skrzypo987) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:56]: You can make the maps above nullable and just check whether they are null.
No change requested

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 4]
```diff
@@ -386,12 +387,19 @@ public static ReaderPageSource createPageSource(
...
         for (List<String> path : descriptorsByPath.keySet()) {
-            paths.add(ColumnPath.get(path.toArray(new String[0])));
+            columnsReadPaths.add(ColumnPath.get(path.toArray(new String[0])));
         }
 
-        return Optional.of(TrinoColumnIndexStore.create(dataSource, blockMetadata, paths));
+        Map<ColumnDescriptor, Domain> parquetDomains = parquetTupleDomain.getDomains()
+                .orElseThrow(() -> new IllegalStateException("Predicate other than none should have domains"));
+        Set<ColumnPath> columnsFilteredPaths = new HashSet<>(parquetDomains.size());
+        for (ColumnDescriptor column : parquetDomains.keySet()) {
```
Comment (skrzypo987) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:398]: This will probably more readable when done using sterams

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java [hunk 5]
```diff
@@ -24,167 +25,202 @@
...
-        private ColumnIndex readColumnIndex(ColumnChunkMetaData meta) throws IOException
-        {
-            IndexReference index = meta.getColumnIndexReference();
-            if (index == null) {
-                return null;
+        if (offsetIndexEnd == -1) {
+            return;
+        }
+        verify(offsetIndexStart < Long.MAX_VALUE);
+        verify(offsetIndexEnd > 0);
```
Comment (skrzypo987) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:183]: verify that end >= start as well

**Review by skrzypo987 — APPROVED** [review]
LGTM % minor comments.
Do you have any reasonable assumptions about the gain that we can aciheve here?

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java (hunk 3 above)
Comment (raunaqmorarka) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:57]: that would work too, but the current way feels safer and easier to read

**Comment by raunaqmorarka** [comment]
> LGTM % minor comments. Do you have any reasonable assumptions about the gain that we can aciheve here?

I observed upto 10% gain in some queries where page filtering wasn't very effective but we were reading indexes for multiple columns one by one. But it depends a lot on the latency of FS calls, so can't make any definitive claims. It is similar to the max-merge-distance optimizations in orc and parquet readers and prefetch of file tail done for reading parquet footer.

**Comment by skrzypo987** [comment]
> I observed upto 10% gain in some queries where page filtering wasn't very effective but we were reading indexes for multiple columns one by one. But it depends a lot on the latency of FS calls, so can't make any definitive claims. It is similar to the max-merge-distance optimizations in orc and parquet readers and prefetch of file tail done for reading parquet footer.

I understand there is no regressionanywhere ? Except maybe for a case when FS latency is very small and bandwidth is huge, which is never the case.

**Comment by JamesRTaylor** [comment]
This looks like a good improvement. Any idea why the overall performance improvement when a selective column index is used isn't much higher? Or has this improved significantly with the couple of bug fixes you've pursued?

**Comment by martint** [comment]
Instead of custom logic for reading the indexes in batch, another option would be to collect all the offset+length pairs in `createPageSource` and use `planRead` to perform the I/O -- that will use the same logic as elsewhere for merging segments if they are close, etc.

**Comment by raunaqmorarka** [comment]
> Instead of custom logic for reading the indexes in batch, another option would be to collect all the offset+length pairs in `createPageSource` and use `planRead` to perform the I/O -- that will use the same logic as elsewhere for merging segments if they are close, etc.

I've updated the code now to reuse `planRead`. The batches for column and offset index are still different because we can skip reading offset indexes if row group was pruned using row group min/max or column index or dictionary. We can also skip offset index when row ranges from column index cover entire row group.

**Comment by martint** [comment]
> The batches for column and offset index are still different because we can skip reading offset indexes if row group was pruned using row group min/max or column index or dictionary.

I haven't looked to closely, but I get a sense that the code in createPageSource + getColumnIndexStore could be refactored so that is the case.

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java [hunk 6]
```diff
@@ -24,167 +31,200 @@
...
-        {
-            return null;
+        Multimap<ColumnIndexMetadata, ChunkReader> chunkReaders = dataSource.planRead(ranges);
+        for (ColumnIndexMetadata column : columnIndexMetadata) {
+            ChunkReader chunkReader = getOnlyElement(chunkReaders.get(column));
+            indexDeserializer.accept(chunkReader.read(), column);
+            chunkReader.free();
         }
-    };
+    }
```
[the remark this answers is not in the document]
Comment (martint) [reply lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:180]: Also, you want to make sure *all* chunk readers are freed. The current code guarantees that, but it's indirect and a harder to reason about. You can rewrite it as follows to be more explicit:

```java
private static <T> Map<ColumnPath, T> loadIndexes(ParquetDataSource dataSource, List<ColumnIndexMetadata> indexMetadata, BiFunction<Slice, ColumnIndexMetadata, T> deserializer)
{
    // Merge multiple small reads of the file for indexes stored close to each other
    ListMultimap<ColumnPath, DiskRange> ranges = ArrayListMultimap.create(indexMetadata.size(), 1);
    for (ColumnIndexMetadata column : indexMetadata) {
        ranges.put(column.getPath(), column.getDiskRange());
    }

    Multimap<ColumnPath, ChunkReader> chunkReaders = dataSource.planRead(ranges);
    try {
        return indexMetadata.stream()
                .collect(Collectors.toMap(
                        ColumnIndexMetadata::getPath,
                        column -> deserializer.apply(getOnlyElement(chunkReaders.get(column.getPath())).read(), column)));
    }
    finally {
        chunkReaders.values().stream()
                .forEach(ChunkReader::free);
    }
}
```

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java [hunk 7]
```diff
@@ -368,9 +368,10 @@ public static ReaderPageSource createPageSource(
             ParquetDataSource dataSource,
             BlockMetaData blockMetadata,
             Map<List<String>, RichColumnDescriptor> descriptorsByPath,
+            TupleDomain<ColumnDescriptor> parquetTupleDomain,
             ParquetReaderOptions options)
     {
-        if (!options.isUseColumnIndex()) {
+        if (!options.isUseColumnIndex() || parquetTupleDomain.isAll() || parquetTupleDomain.isNone()) {
```
Comment (martint) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/parquet/ParquetPageSourceFactory.java:374]: This seems like an unrelated improvement. Move to a separate commit.

**Review by martint — CHANGES_REQUESTED** [review]

File: lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java [hunk 8]
```diff
@@ -24,167 +31,166 @@
...
+            columnIndexStore = loadIndexes(dataSource, columnIndexReferences, (buffer, columnMetadata) -> {
                 try {
-                    columnIndex = readColumnIndex(meta);
+                    return ParquetMetadataConverter.fromParquetColumnIndex(columnMetadata.getPrimitiveType(), Util.readColumnIndex(buffer.getInput()));
                 }
                 catch (IOException e) {
                     // If the I/O issue still stands it will fail the reading later;
                     // otherwise we fail the filtering only with a missing column index.
+                    LOG.warn(e, "Unable to read column index for column %s", columnMetadata.getPath());
+                    return null;
```
Comment (martint) [inline lib/trino-parquet/src/main/java/io/trino/parquet/reader/TrinoColumnIndexStore.java:116]: There’s no actual I/O here. A failure in this context indicates the column index is corrupted, so we should fail with a proper exception.
